src/main/java/uk/ac/ic/wlgitbridge/bridge/db/postgres/PostgresDBStore.java

```diff
@@ -0,0 +1,298 @@
+      List<String> result = new ArrayList<String>();
+      while (rs.next()) {
+        result.add(rs.getString(1));
+      }
+      rs.close();
+      statement.close();
+      return result;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void setLatestVersionForProject(String project, int versionID) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "INSERT INTO "
+          + "projects (name, version_id, last_accessed) "
+          + "VALUES (?, ?, now()) "
+          + "ON CONFLICT (name) DO UPDATE "
+          + "SET version_id = EXCLUDED.version_id, last_accessed = now();\n"
+      );
+      statement.setString(1, project);
+      statement.setInt(2, versionID);
+      statement.executeUpdate();
+      statement.close();
+      return;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public int getLatestVersionForProject(String project) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "SELECT version_id from projects where name = ?;"
+      );
+      statement.setString(1, project);
+      ResultSet rs = statement.executeQuery();
+      if (rs.next()) {
+        int versionId = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return versionId;
+      } else {
+        return 0;
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void addURLIndexForProject(String projectName, String url, String path) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "INSERT INTO url_index_store(" +
+          "project_name, " +
+          "url, " +
+          "path" +
+          ") VALUES " +
+          "(?, ?, ?)" +
+          "ON CONFLICT (project_name,path) DO UPDATE " +
+          "SET url = EXCLUDED.url;\n"
+        // TODO: check this is actually correct, with the two indexes
+      );
+      statement.setString(1, projectName);
+      statement.setString(2, url);
+      statement.setString(3, path);
+      statement.executeUpdate();
+      statement.close();
+      return;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void deleteFilesForProject(String project, String... files) {
+    if (files.length == 0) {
+      return; // TODO: check this is correct
+    }
+    try {
+      StringBuilder queryBuilder = new StringBuilder();
+      queryBuilder.append("DELETE FROM url_index_store ");
+      queryBuilder.append("WHERE project_name = ? AND path IN (");
+      for (int i = 0; i < files.length; i++) {
```

Comment:
I'm not sure what the Java standard library gives you, but if you have a map on a string list you could say this more cleanly with something along the lines of this javascript equivalent?
```JavaScript
queryBuilder.append(files.map(() => "?").join(", "))
```

I've definitely written this same loop before though in C. ;-)